src/commands/find-buf.ts

```diff
@@ -0,0 +1,143 @@
+            configPath
+          );
+        }
+
+        if (configVersion) {
+          log.warn(
+            "Both 'buf.commandLine.path' and 'buf.commandLine.version' are set. Using 'buf.commandLine.path'."
+          );
+        }
+
+        try {
+          log.info(`Buf path set to '${configPath}'.`);
+          bufCtx.buf = await BufVersion.fromPath(configPath);
+
+          if (bufCtx.buf) {
+            log.info(
+              `Using '${bufCtx.buf.path}', version: ${bufCtx.buf.version}.`
+            );
+          }
+        } catch (e) {
+          log.error(
+            `Error loading buf from path '${configPath}': ${unwrapError(e)}`
+          );
+          bufCtx.buf = undefined;
+        }
+
+        return;
+      }
+
+      if (configVersion) {
+        try {
+          log.info(
+            `Buf version set to '${configVersion}'. Looking for it in the extension storage...`
+          );
+          bufCtx.buf = await findBufInStorage(ctx, configVersion);
+
+          if (bufCtx.buf) {
+            log.info(
+              `Using '${bufCtx.buf.path}', version: ${bufCtx.buf.version}.`
+            );
+          }
+        } catch (e) {
+          log.error(`Error finding an installed buf: ${unwrapError(e)}`);
+          bufCtx.buf = undefined;
+        }
+
+        return;
+      }
+
+      // If we didn't get a valid buf binary from config...
+      log.info("Looking for buf on the path...");
+
+      // If we didn't find it in storage...
+      try {
+        bufCtx.buf = await findBufInPath();
+
+        if (bufCtx.buf) {
+          log.info(
+            `Using '${bufCtx.buf.path}', version: ${bufCtx.buf.version}.`
+          );
+          return;
+        }
+      } catch (e) {
+        log.error(`Error finding an installed buf: ${unwrapError(e)}`);
+      }
+
+      // If we made it to this point, we found nothing, throw an error.
+      log.error(
+        "Buf is not installed or no valid binary could be found. Please install Buf."
+      );
+    };
+  }
+);
+
+const findBufInPath = async (): Promise<BufVersion | undefined> => {
+  const bufPath = await which(bufFilename, { nothrow: true });
+
+  if (bufPath) {
+    return BufVersion.fromPath(bufPath);
+  }
+
+  return undefined;
+};
+
+const findBufInStorage = async (
+  ctx: vscode.ExtensionContext,
+  version: string
+): Promise<BufVersion | undefined> => {
+  const files = await fs.promises.readdir(ctx.globalStorageUri.fsPath);
+
+  // If the user has not set a specific version, we do nothing.
+  if (!version) {
+    return undefined;
+  }
+
+  let found: string | undefined;
+
+  if (version === "latest") {
+    found = files
+      .filter((f) => f.startsWith("v"))
+      .sort()
+      .at(-1);
+  } else {
+    found = files.find((f) => f.localeCompare(version) === 0);
+  }
+
+  if (found) {
+    return BufVersion.fromPath(
+      path.join(ctx.globalStorageUri.fsPath, found, bufFilename)
+    );
+  }
+
+  return undefined;
```

Comment:
I think we could simplify (but also add a async request) with something like:

```suggestion
  if (version === "latest") {
    const latestRelease = await github.getRelease("latest");
    version = latestRelease.tag_name;
  }

  if (!files.includes(version)) {
    return undefined
  }
  return BufVersion.fromPath(
    path.join(ctx.globalStorageUri.fsPath, version, bufFilename)
  );
```

Reply:
Thanks @paul-sachs - took a slightly different approach but same outcome.